File: android/core/common/src/main/java/com/akaar/core/common/UiState.kt
package com.akaar.core.common

/**
 * Every screen in Akaar must handle six states. Making them a sealed hierarchy
 * rather than a set of booleans means an unhandled state is a compile error
 * instead of something discovered during the demo.
 *
 * The design system supplies a scaffold for each, so honouring this is cheap.
 */
sealed interface UiState<out T> {

    /** Work in flight and the shape of the result is known: show a skeleton. */
    data object Loading : UiState<Nothing>

    data class Content<out T>(val data: T) : UiState<T>

    /**
     * Succeeded, nothing to show. Always names the action that would create
     * something - an empty screen with no next step is a dead end.
     */
    data class Empty(
        val title: String,
        val body: String? = null,
        val actionLabel: String? = null,
    ) : UiState<Nothing>

    /**
     * Something failed. [cause] is for logs; [message] is plain language for the
     * user, and [audioPrompt] is what gets read aloud in seller flows.
     */
    data class Error(
        val message: String,
        val cause: Throwable? = null,
        val retryable: Boolean = true,
        val audioPrompt: String? = null,
    ) : UiState<Nothing>

    /**
     * Distinct from Error on purpose: "you are offline" and "this failed" call
     * for different words and different next steps. [safeToContinue] tells the
     * user what still works without a network.
     */
    data class Offline(
        val safeToContinue: String? = null,
    ) : UiState<Nothing>

    /** Explains why the permission is needed, and offers settings after a denial. */
    data class PermissionDenied(
        val permission: String,
        val rationale: String,
        val permanentlyDenied: Boolean = false,
    ) : UiState<Nothing>
}

inline fun <T, R> UiState<T>.map(transform: (T) -> R): UiState<R> = when (this) {
    is UiState.Content -> UiState.Content(transform(data))
    is UiState.Loading -> this
    is UiState.Empty -> this
    is UiState.Error -> this
    is UiState.Offline -> this
    is UiState.PermissionDenied -> this
}

val <T> UiState<T>.contentOrNull: T? get() = (this as? UiState.Content)?.data

File: mobile/packages/core-common/src/ui-state.ts
import { assertNever } from "./result";

/**
 * Port of android/core/common/.../UiState.kt.
 *
 * Every screen in Akaar must handle six states. In Kotlin, making them a
 * `sealed interface` turns an unhandled state into a compile error instead
 * of something discovered during the demo. A TypeScript discriminated union
 * gets the same property: `switch` on `state.kind` plus {@link assertNever}
 * in the `default` branch fails the *build*, not just the test, on a missing
 * case - see `mapUiState` below.
 *
 * The design system supplies a scaffold for each, so honouring this is cheap
 * (see packages/design-system/src/components/StateHost.tsx).
 */
export type UiState<T> =
  /** Work in flight and the shape of the result is known: show a skeleton. */
  | { kind: "loading" }
  | { kind: "content"; data: T }
  /**
   * Succeeded, nothing to show. Always names the action that would create
   * something - an empty screen with no next step is a dead end.
   */
  | { kind: "empty"; title: string; body?: string; actionLabel?: string }
  /**
   * Something failed. `cause` is for logs; `message` is plain language for
   * the user, and `audioPrompt` is what gets read aloud in seller flows.
   */
  | {
      kind: "error";
      message: string;
      cause?: unknown;
      retryable?: boolean;
      audioPrompt?: string;
    }
  /**
   * Distinct from Error on purpose: "you are offline" and "this failed" call
   * for different words and different next steps. `safeToContinue` tells the
   * user what still works without a network.
   */
  | { kind: "offline"; safeToContinue?: string }
  /** Explains why the permission is needed, and offers settings after a denial. */
  | {
      kind: "permissionDenied";
      permission: string;
      rationale: string;
      permanentlyDenied?: boolean;
    };

export const UiState = {
  loading: <T = never>(): UiState<T> => ({ kind: "loading" }),
  content: <T>(data: T): UiState<T> => ({ kind: "content", data }),
  empty: <T = never>(title: string, body?: string, actionLabel?: string): UiState<T> => ({
    kind: "empty",
    title,
    body,
    actionLabel,
  }),
  error: <T = never>(
    message: string,
    opts: { cause?: unknown; retryable?: boolean; audioPrompt?: string } = {},
  ): UiState<T> => ({ kind: "error", message, retryable: opts.retryable ?? true, ...opts }),
  offline: <T = never>(safeToContinue?: string): UiState<T> => ({ kind: "offline", safeToContinue }),
  permissionDenied: <T = never>(
    permission: string,
    rationale: string,
    permanentlyDenied = false,
  ): UiState<T> => ({ kind: "permissionDenied", permission, rationale, permanentlyDenied }),
};

export function mapUiState<T, R>(state: UiState<T>, transform: (data: T) => R): UiState<R> {
  switch (state.kind) {
    case "content":
      return { ...state, data: transform(state.data) };
    case "loading":
    case "empty":
    case "error":
    case "offline":
    case "permissionDenied":
      return state;
    default:
      return assertNever(state);
  }
}

export function contentOrNull<T>(state: UiState<T>): T | null {
  return state.kind === "content" ? state.data : null;
}
